src/Input/Input.js

```diff
@@ -0,0 +1,272 @@
+        transition: theme.transitions.create(
+          'transform',
+          '200ms',
+          null,
+          easing.easeOut,
+        ),
+      },
+      '&$focused:after': {
+        transform: 'scaleX(1)',
+      },
+    },
+    focused: {},
+    error: {
+      '&:after': {
+        backgroundColor: theme.palette.error[500],
+        transform: 'scaleX(1)', // error is always underlined in red
+      },
+    },
+    input: {
+      font: 'inherit',
+      padding: '6px 0',
+      border: 0,
+      display: 'block',
+      verticalAlign: 'middle',
+      whiteSpace: 'normal',
+      background: 'none',
+      lineHeight: 1,
+      appearance: 'textfield', // Improve type search style.
+      color: theme.palette.text.primary,
+      '&:focus': {
+        outline: 0,
+      },
+      '&::-webkit-search-decoration': { // Remove the padding when type=search.
+        appearance: 'none',
+      },
+    },
+    disabled: {
+      color: theme.palette.text.disabled,
+      cursor: 'not-allowed',
+    },
+    underline: {
+      borderBottom: `1px solid ${theme.palette.text.divider}`,
+      '&$disabled': {
+        borderBottomStyle: 'dotted',
+      },
+    },
+  };
+});
+
+/**
+ * Input
+ */
+export default class Input extends Component {
+  static muiName = 'Input';
+
+  static propTypes = {
+    /**
+     * The CSS class name of the wrapper element.
+     */
+    className: PropTypes.string,
+    /**
+     * The element or component used for the root node.
+     */
+    component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
+    disabled: PropTypes.bool,
+    error: PropTypes.bool,
+    /**
+     * The CSS class name of the input element.
+     */
+    inputClassName: PropTypes.string,
+    /**
+     * @ignore
+     */
+    onBlur: PropTypes.func,
+    /**
+     * @ignore
+     */
+    onChange: PropTypes.func,
+    /**
+     * @ignore
+     */
+    onClean: PropTypes.func,
+    /**
+     * @ignore
+     */
+    onDirty: PropTypes.func,
+    /**
+     * @ignore
+     */
+    onFocus: PropTypes.func,
+    /**
+     * Input type.
+     */
+    type: PropTypes.string,
+    /**
+     * If set to true, the input will have an underline.
+     */
+    underline: PropTypes.bool,
+    /**
+     * The input value, required for a controlled component.
+     */
+    value: PropTypes.string,
+  };
+
+  static defaultProps = {
+    component: 'input',
+    type: 'text',
+    underline: true,
+  };
+
+  static contextTypes = {
+    formControl: PropTypes.object,
+    styleManager: PropTypes.object.isRequired,
+  };
+
+  state = {
+    focused: false,
+  };
+
+  componentWillMount() {
+    if (this.isControlled()) {
+      this.checkDirty(this.props);
+    }
+  }
+
+  componentDidMount() {
+    if (!this.isControlled()) {
+      this.checkDirty(this.input);
+    }
+  }
+
+  componentWillUpdate(nextProps) {
+    if (this.isControlled()) {
+      this.checkDirty(nextProps);
+    }
+  }
+
+  // Holds the input reference
+  input = undefined;
```

Comment:
@oliviertassinari I don't have a specific reason. We can change these to `null` if it is consistent with React.

Also, another thing regarding refs that I was wondering (not applicable here). Do you think that if we know the ref is a component instance rather than an HTML DOM element, do you think that we should uppercase the first letter?